src/toil/test/src/jobCacheTest.py

```diff
@@ -1125,6 +1126,106 @@ def testBatchSystemCleanupCanHandleWorkerDeaths(self):
+            nonLocalFile = os.path.join(workdir, str(uuid4()))
+            # The first file has to exist and be empty
+            file1 = open(nonLocalFile, 'w').close()
+            assert os.path.exists(nonLocalFile)
+            root = Job()
+            A = Job.wrapJobFn(self._testOrphanedFilesAreDeletedFromJobStore_A,
+                              nonLocalFile=nonLocalFile)
+            B = Job.wrapJobFn(self._testOrphanedFilesAreDeletedFromJobStore_B,
+                              nonLocalFile=nonLocalFile)
+            C = Job.wrapJobFn(self._testOrphanedFilesAreDeletedFromJobStore_C,
+                              nonLocalFile=nonLocalFile)
+            root.addChild(A)
+            root.addChild(B)
+            B.addChild(C)
+            try:
+                Job.Runner.startToil(root, self.options)
+            except FailedJobsException as err:
+                self.assertEqual(err.numberOfFailedJobs, 2) # root and A
+            assert not os.path.exists(nonLocalFile)
+
+        @staticmethod
+        def _testOrphanedFilesAreDeletedFromJobStore_A(job, nonLocalFile):
+            """
+            Write a file to the jobstore and write the jobstore file ID and self pid to
+            nonLocalFile. Then die.
+
+            :param job: Job
+            :param nonLocalFile: A file outside the scope of the working directory that can persist
+                   across jobs.
+            :returns: None
+            """
+            jsID = job.fileStore.writeGlobalFile(job.fileStore.getLocalTempFile())
+            with open(nonLocalFile, 'w') as fH:
+                dill.dump({'jsID': jsID, 'pid': os.getpid()}, fH)
+            os.kill(os.getpid(), signal.SIGKILL)
+
+        @staticmethod
+        def _testOrphanedFilesAreDeletedFromJobStore_B(job, nonLocalFile):
+            """
+            Wait for nonLocalFile to get populated. Ensure that the file exists in teh jobstore.
+            Then spawn the child that will check for non-existence of the jobstore file after the
+            scheduled cleanup.
+
+            :param job: Job
+            :param nonLocalFile: A file outside the scope of the working directory that can persist
+                   across jobs.
+            :returns: None
+            """
+            attempt = 1
+            while attempt <= 10:
+                try:
+                    with open(nonLocalFile, 'r') as fH:
+                        jobAInfo = dill.load(fH)
+                except EOFError:
+                    time.sleep(3)
+                    attempt += 1
+                else:
+                    while job.fileStore.HarbingerFile._pidExists(jobAInfo['pid']):
+                        # This should never infinite loop.
+                        time.sleep(1)
+                    break
+            else:
+                raise RuntimeError('Timed out waiting for nonLocalFile.')
+            # The file should exist at this point.
+            job.fileStore.readGlobalFile(jobAInfo['jsID'])
```

Comment:
Isn't this racy? The file could be deleted before you get to this. If you argue that its not racy due to some implementation detail of the deferred mechanism, I would say that the test shouldn't rely on that. 

I would actually be satisfied if the existence of the file is asserted in A and the non-existence asserted in C.